File: backend/tests/test_project_root.py
#!/usr/bin/env python3
"""
🦊 Project Root Utility Tests for Python
========================================

Comprehensive tests to ensure the Python project root utility
behaves consistently with Shell and TypeScript versions.

Author: Reynard Development Team
Version: 1.0.0
"""

import os

# Add the backend directory to the Python path
import sys
import tempfile
import unittest
from pathlib import Path
from unittest.mock import patch

sys.path.insert(0, str(Path(__file__).parent.parent))

from app.core.project_root import (
    BACKEND_DIR,
    E2E_DIR,
    EXAMPLES_DIR,
    EXPERIMENTAL_DIR,
    PAPERS_DIRECTORY,
    PROJECT_ROOT,
    get_backend_dir,
    get_e2e_dir,
    get_examples_dir,
    get_experimental_dir,
    get_papers_directory,
    get_project_path,
    get_project_root,
)


class TestProjectRootUtility(unittest.TestCase):
    """Test suite for the Python project root utility."""

    def setUp(self):
        """Set up test environment."""
        self.expected_project_root = "/home/kade/runeset/reynard"
        self.expected_backend_dir = "/home/kade/runeset/reynard/backend"
        self.expected_e2e_dir = "/home/kade/runeset/reynard/e2e"
        self.expected_examples_dir = "/home/kade/runeset/reynard/examples"
        self.expected_experimental_dir = "/home/kade/runeset/reynard/experimental"
        self.expected_papers_dir = "/home/kade/runeset/reynard/backend/data/papers"

    def test_get_project_root_default(self):
        """Test default project root detection."""
        result = get_project_root()
        self.assertEqual(str(result), self.expected_project_root)
        self.assertIsInstance(result, Path)

    def test_get_project_root_with_env_var(self):
        """Test project root with environment variable override."""
        with tempfile.TemporaryDirectory() as temp_dir:
            with patch.dict(os.environ, {'REYNARD_PROJECT_ROOT': temp_dir}):
                result = get_project_root()
                self.assertEqual(str(result), temp_dir)

    def test_get_project_path(self):
        """Test getting paths relative to project root."""
        result = get_project_path("backend")
        self.assertEqual(str(result), self.expected_backend_dir)

        result = get_project_path("e2e")
        self.assertEqual(str(result), self.expected_e2e_dir)

    def test_get_backend_dir(self):
        """Test backend directory function."""
        result = get_backend_dir()
        self.assertEqual(str(result), self.expected_backend_dir)
        self.assertIsInstance(result, Path)

    def test_get_e2e_dir(self):
        """Test E2E directory function."""
        result = get_e2e_dir()
        self.assertEqual(str(result), self.expected_e2e_dir)
        self.assertIsInstance(result, Path)

    def test_get_examples_dir(self):
        """Test examples directory function."""
        result = get_examples_dir()
        self.assertEqual(str(result), self.expected_examples_dir)
        self.assertIsInstance(result, Path)

    def test_get_experimental_dir(self):
        """Test experimental directory function."""
        result = get_experimental_dir()
        self.assertEqual(str(result), self.expected_experimental_dir)
        self.assertIsInstance(result, Path)

    def test_get_papers_directory(self):
        """Test papers directory function."""
        result = get_papers_directory()
        self.assertEqual(str(result), self.expected_papers_dir)
        self.assertIsInstance(result, Path)

    def test_constant_exports(self):
        """Test that constants are exported correctly."""
        self.assertEqual(str(PROJECT_ROOT), self.expected_project_root)
        self.assertEqual(str(BACKEND_DIR), self.expected_backend_dir)
        self.assertEqual(str(E2E_DIR), self.expected_e2e_dir)
        self.assertEqual(str(EXAMPLES_DIR), self.expected_examples_dir)
        self.assertEqual(str(EXPERIMENTAL_DIR), self.expected_experimental_dir)
        self.assertEqual(str(PAPERS_DIRECTORY), self.expected_papers_dir)

    def test_directory_existence(self):
        """Test that all directories exist."""
        self.assertTrue(
            PROJECT_ROOT.exists(), f"Project root should exist: {PROJECT_ROOT}"
        )
        self.assertTrue(
            BACKEND_DIR.exists(), f"Backend directory should exist: {BACKEND_DIR}"
        )
        self.assertTrue(E2E_DIR.exists(), f"E2E directory should exist: {E2E_DIR}")
        self.assertTrue(
            EXAMPLES_DIR.exists(), f"Examples directory should exist: {EXAMPLES_DIR}"
        )
        self.assertTrue(
            EXPERIMENTAL_DIR.exists(),
            f"Experimental directory should exist: {EXPERIMENTAL_DIR}",
        )

    def test_project_structure_validation(self):
        """Test that key project structure elements exist."""
        # Test that package.json exists in project root
        package_json = PROJECT_ROOT / "package.json"
        self.assertTrue(
            package_json.exists(), f"package.json should exist: {package_json}"
        )

        # Test that package.json contains "reynard"
        if package_json.exists():
            content = package_json.read_text()
            self.assertIn(
                "reynard", content.lower(), "package.json should contain 'reynard'"
            )

    def test_consistency_with_functions(self):
        """Test that functions return the same values as constants."""
        self.assertEqual(get_project_root(), PROJECT_ROOT)
        self.assertEqual(get_backend_dir(), BACKEND_DIR)
        self.assertEqual(get_e2e_dir(), E2E_DIR)
        self.assertEqual(get_examples_dir(), EXAMPLES_DIR)
        self.assertEqual(get_experimental_dir(), EXPERIMENTAL_DIR)
        self.assertEqual(get_papers_directory(), PAPERS_DIRECTORY)

    def test_path_consistency(self):
        """Test that all paths are consistent with project root."""
        self.assertEqual(BACKEND_DIR, PROJECT_ROOT / "backend")
        self.assertEqual(E2E_DIR, PROJECT_ROOT / "e2e")
        self.assertEqual(EXAMPLES_DIR, PROJECT_ROOT / "examples")
        self.assertEqual(EXPERIMENTAL_DIR, PROJECT_ROOT / "experimental")
        self.assertEqual(PAPERS_DIRECTORY, PROJECT_ROOT / "backend" / "data" / "papers")

    def test_environment_variable_priority(self):
        """Test that environment variable takes priority over default detection."""
        with tempfile.TemporaryDirectory() as temp_dir:
            # Create a fake package.json to test environment variable priority
            fake_package_json = Path(temp_dir) / "package.json"
            fake_package_json.write_text('{"name": "fake-reynard"}')

            with patch.dict(os.environ, {'REYNARD_PROJECT_ROOT': temp_dir}):
                result = get_project_root()
                self.assertEqual(str(result), temp_dir)

    def test_fallback_behavior(self):
        """Test fallback behavior when environment variable is invalid."""
        with patch.dict(os.environ, {'REYNARD_PROJECT_ROOT': '/nonexistent/path'}):
            result = get_project_root()
            # Should fall back to default detection
            self.assertEqual(str(result), self.expected_project_root)


class TestProjectRootConsistency(unittest.TestCase):
    """Test consistency between different project root utilities."""

    def test_expected_values(self):
        """Test that all utilities return the same expected values."""
        expected_values = {
            'project_root': '/home/kade/runeset/reynard',
            'backend_dir': '/home/kade/runeset/reynard/backend',
            'e2e_dir': '/home/kade/runeset/reynard/e2e',
            'examples_dir': '/home/kade/runeset/reynard/examples',
            'experimental_dir': '/home/kade/runeset/reynard/experimental',
            'papers_dir': '/home/kade/runeset/reynard/backend/data/papers',
        }

        # Test Python values
        python_values = {
            'project_root': str(get_project_root()),
            'backend_dir': str(get_backend_dir()),
            'e2e_dir': str(get_e2e_dir()),
            'examples_dir': str(get_examples_dir()),
            'experimental_dir': str(get_experimental_dir()),
            'papers_dir': str(get_papers_directory()),
        }

        for key, expected in expected_values.items():
            with self.subTest(key=key):
                self.assertEqual(
                    python_values[key],
                    expected,
                    f"Python {key} should match expected value",
                )


if __name__ == '__main__':
    # Run tests with verbose output
    unittest.main(verbosity=2)

File: scripts/utils/test-project-root.ts
#!/usr/bin/env node
/**
 * @fileoverview Project Root Utility Tests for Node.js/TypeScript
 *
 * 🦊 *whiskers twitch with testing precision* Comprehensive tests to ensure
 * the TypeScript project root utility behaves consistently with Shell and Python versions.
 *
 * @author Strategic-Fox-42 (Reynard Fox Specialist)
 * @since 1.0.0
 */

import { existsSync } from "fs";
import { join } from "path";
import {
  getProjectRoot,
  getProjectPath,
  getE2EDir,
  getBackendDir,
  getExamplesDir,
  getPackagesDir,
  PROJECT_ROOT,
  E2E_DIR,
  BACKEND_DIR,
  EXAMPLES_DIR,
  PACKAGES_DIR,
} from "./project-root.js";

// Test counters
let testsRun = 0;
let testsPassed = 0;
let testsFailed = 0;

// Colors for output
const colors = {
  red: "\x1b[31m",
  green: "\x1b[32m",
  blue: "\x1b[34m",
  yellow: "\x1b[33m",
  reset: "\x1b[0m",
};

/**
 * Run a single test
 * @param testName - Name of the test
 * @param testFunction - Function that returns the test result
 * @param expectedResult - Expected result
 */
function runTest(testName: string, testFunction: () => any, expectedResult: any): void {
  testsRun++;
  console.log(`${colors.blue}🧪 Running test: ${testName}${colors.reset}`);

  try {
    const actualResult = testFunction();

    if (actualResult === expectedResult) {
      console.log(`  ${colors.green}✅ PASS${colors.reset}: ${testName}`);
      console.log(`    Expected: ${expectedResult}`);
      console.log(`    Got:      ${actualResult}`);
      testsPassed++;
    } else {
      console.log(`  ${colors.red}❌ FAIL${colors.reset}: ${testName}`);
      console.log(`    Expected: ${expectedResult}`);
      console.log(`    Got:      ${actualResult}`);
      testsFailed++;
    }
  } catch (error) {
    console.log(`  ${colors.red}❌ FAIL${colors.reset}: ${testName} (error: ${(error as Error).message})`);
    testsFailed++;
  }
  console.log();
}

/**
 * Test default behavior
 */
function testDefaultBehavior(): void {
  console.log(`${colors.yellow}🔧 Testing default behavior...${colors.reset}`);

  const expectedProjectRoot = "/home/kade/runeset/reynard";

  // Test main project root
  runTest("Default project root", () => getProjectRoot(), expectedProjectRoot);
  runTest("PROJECT_ROOT constant", () => PROJECT_ROOT, expectedProjectRoot);

  // Test convenience functions
  runTest("getE2EDir function", () => getE2EDir(), "/home/kade/runeset/reynard/e2e");
  runTest("getBackendDir function", () => getBackendDir(), "/home/kade/runeset/reynard/backend");
  runTest("getExamplesDir function", () => getExamplesDir(), "/home/kade/runeset/reynard/examples");
  runTest("getPackagesDir function", () => getPackagesDir(), "/home/kade/runeset/reynard/packages");

  // Test constants
  runTest("E2E_DIR constant", () => E2E_DIR, "/home/kade/runeset/reynard/e2e");
  runTest("BACKEND_DIR constant", () => BACKEND_DIR, "/home/kade/runeset/reynard/backend");
  runTest("EXAMPLES_DIR constant", () => EXAMPLES_DIR, "/home/kade/runeset/reynard/examples");
  runTest("PACKAGES_DIR constant", () => PACKAGES_DIR, "/home/kade/runeset/reynard/packages");
}

/**
 * Test getProjectPath function
 */
function testGetProjectPath(): void {
  console.log(`${colors.yellow}🔧 Testing getProjectPath function...${colors.reset}`);

  runTest("getProjectPath with backend", () => getProjectPath("backend"), "/home/kade/runeset/reynard/backend");
  runTest("getProjectPath with e2e", () => getProjectPath("e2e"), "/home/kade/runeset/reynard/e2e");
  runTest("getProjectPath with examples", () => getProjectPath("examples"), "/home/kade/runeset/reynard/examples");
  runTest("getProjectPath with packages", () => getProjectPath("packages"), "/home/kade/runeset/reynard/packages");
  runTest(
    "getProjectPath with nested path",
    () => getProjectPath("backend/data"),
    "/home/kade/runeset/reynard/backend/data"
  );
}

/**
 * Test directory existence
 */
function testDirectoryExistence(): void {
  console.log(`${colors.yellow}🔧 Testing directory existence...${colors.reset}`);

  runTest("Project root exists", () => (existsSync(PROJECT_ROOT) ? "exists" : "missing"), "exists");
  runTest("Backend directory exists", () => (existsSync(BACKEND_DIR) ? "exists" : "missing"), "exists");
  runTest("E2E directory exists", () => (existsSync(E2E_DIR) ? "exists" : "missing"), "exists");
  runTest("Examples directory exists", () => (existsSync(EXAMPLES_DIR) ? "exists" : "missing"), "exists");
  runTest("Packages directory exists", () => (existsSync(PACKAGES_DIR) ? "exists" : "missing"), "exists");
}

/**
 * Test project structure validation
 */
function testProjectStructure(): void {
  console.log(`${colors.yellow}🔧 Testing project structure validation...${colors.reset}`);

  // Test that package.json exists in project root
  const packageJsonPath = join(PROJECT_ROOT, "package.json");
  runTest("Package.json exists", () => (existsSync(packageJsonPath) ? "exists" : "missing"), "exists");

  // Test that key directories exist
  runTest(
    "Backend directory exists in project root",
    () => (existsSync(join(PROJECT_ROOT, "backend")) ? "exists" : "missing"),
    "exists"
  );
  runTest(
    "E2E directory exists in project root",
    () => (existsSync(join(PROJECT_ROOT, "e2e")) ? "exists" : "missing"),
    "exists"
  );
  runTest(
    "Examples directory exists in project root",
    () => (existsSync(join(PROJECT_ROOT, "examples")) ? "exists" : "missing"),
    "exists"
  );
  runTest(
    "Packages directory exists in project root",
    () => (existsSync(join(PROJECT_ROOT, "packages")) ? "exists" : "missing"),
    "exists"
  );
}

/**
 * Test consistency between functions and constants
 */
function testConsistency(): void {
  console.log(`${colors.yellow}🔧 Testing consistency between functions and constants...${colors.reset}`);

  runTest(
    "getProjectRoot() === PROJECT_ROOT",
    () => (getProjectRoot() === PROJECT_ROOT ? "consistent" : "inconsistent"),
    "consistent"
  );
  runTest("getE2EDir() === E2E_DIR", () => (getE2EDir() === E2E_DIR ? "consistent" : "inconsistent"), "consistent");
  runTest(
    "getBackendDir() === BACKEND_DIR",
    () => (getBackendDir() === BACKEND_DIR ? "consistent" : "inconsistent"),
    "consistent"
  );
  runTest(
    "getExamplesDir() === EXAMPLES_DIR",
    () => (getExamplesDir() === EXAMPLES_DIR ? "consistent" : "inconsistent"),
    "consistent"
  );
  runTest(
    "getPackagesDir() === PACKAGES_DIR",
    () => (getPackagesDir() === PACKAGES_DIR ? "consistent" : "inconsistent"),
    "consistent"
  );
}

/**
 * Test path consistency
 */
function testPathConsistency(): void {
  console.log(`${colors.yellow}🔧 Testing path consistency...${colors.reset}`);

  runTest(
    "BACKEND_DIR === PROJECT_ROOT/backend",
    () => (BACKEND_DIR === join(PROJECT_ROOT, "backend") ? "consistent" : "inconsistent"),
    "consistent"
  );
  runTest(
    "E2E_DIR === PROJECT_ROOT/e2e",
    () => (E2E_DIR === join(PROJECT_ROOT, "e2e") ? "consistent" : "inconsistent"),
    "consistent"
  );
  runTest(
    "EXAMPLES_DIR === PROJECT_ROOT/examples",
    () => (EXAMPLES_DIR === join(PROJECT_ROOT, "examples") ? "consistent" : "inconsistent"),
    "consistent"
  );
  runTest(
    "PACKAGES_DIR === PROJECT_ROOT/packages",
    () => (PACKAGES_DIR === join(PROJECT_ROOT, "packages") ? "consistent" : "inconsistent"),
    "consistent"
  );
}

/**
 * Test expected values consistency
 */
function testExpectedValues(): void {
  console.log(`${colors.yellow}🔧 Testing expected values consistency...${colors.reset}`);

  const expectedValues = {
    project_root: "/home/kade/runeset/reynard",
    backend_dir: "/home/kade/runeset/reynard/backend",
    e2e_dir: "/home/kade/runeset/reynard/e2e",
    examples_dir: "/home/kade/runeset/reynard/examples",
    packages_dir: "/home/kade/runeset/reynard/packages",
  };

  const actualValues = {
    project_root: getProjectRoot(),
    backend_dir: getBackendDir(),
    e2e_dir: getE2EDir(),
    examples_dir: getExamplesDir(),
    packages_dir: getPackagesDir(),
  };

  for (const [key, expected] of Object.entries(expectedValues)) {
    runTest(
      `TypeScript ${key} matches expected`,
      () => (actualValues[key as keyof typeof actualValues] === expected ? "matches" : "mismatch"),
      "matches"
    );
  }
}

/**
 * Test environment variable override (simplified version)
 */
function testEnvironmentVariableOverride(): void {
  console.log(`${colors.yellow}🔧 Testing environment variable override...${colors.reset}`);

  // Save original environment variable
  const originalEnv = process.env.REYNARD_PROJECT_ROOT;

  try {
    // Test with custom environment variable
    const customRoot = "/tmp/test-reynard-root";
    process.env.REYNARD_PROJECT_ROOT = customRoot;

    // Since the module is already loaded, we can't test dynamic import easily
    // Instead, we'll test that the environment variable is set correctly
    runTest(
      "Environment variable is set",
      () => (process.env.REYNARD_PROJECT_ROOT === customRoot ? "set" : "not set"),
      "set"
    );

    // Note: The actual project root won't change because the module is already loaded
    // This is expected behavior for Node.js modules
    console.log(
      `  ${colors.yellow}ℹ️  INFO${colors.reset}: Environment variable override test is limited because the module is already loaded`
    );
    console.log(`    This is expected behavior for Node.js modules`);
    console.log();
  } catch (error) {
    console.log(
      `  ${colors.yellow}⚠️  SKIP${colors.reset}: Environment variable override (${(error as Error).message})`
    );
  } finally {
    // Restore original environment variable
    if (originalEnv !== undefined) {
      process.env.REYNARD_PROJECT_ROOT = originalEnv;
    } else {
      delete process.env.REYNARD_PROJECT_ROOT;
    }
  }
}

/**
 * Main test function
 */
async function main(): Promise<void> {
  console.log(`${colors.blue}🦊 TypeScript Project Root Utility Tests${colors.reset}`);
  console.log(`${colors.blue}========================================${colors.reset}`);
  console.log();

  // Run all test suites
  testDefaultBehavior();
  testGetProjectPath();
  testDirectoryExistence();
  testProjectStructure();
  testConsistency();
  testPathConsistency();
  testExpectedValues();
  testEnvironmentVariableOverride();

  // Print summary
  console.log(`${colors.blue}📊 Test Summary${colors.reset}`);
  console.log(`${colors.blue}===============${colors.reset}`);
  console.log(`Tests run:    ${testsRun}`);
  console.log(`Tests passed: ${colors.green}${testsPassed}${colors.reset}`);
  console.log(`Tests failed: ${colors.red}${testsFailed}${colors.reset}`);

  if (testsFailed === 0) {
    console.log(`\n${colors.green}🎉 All tests passed!${colors.reset}`);
    process.exit(0);
  } else {
    console.log(`\n${colors.red}💥 Some tests failed!${colors.reset}`);
    process.exit(1);
  }
}

// Run tests
main().catch(error => {
  console.error(`${colors.red}💥 Test runner failed: ${(error as Error).message}${colors.reset}`);
  process.exit(1);
});
